pilot/pkg/serviceregistry/kube/controller/ambient/multicluster.go

```diff
@@ -381,15 +446,107 @@ func (a *index) buildGlobalCollections(
 
 		return []networkAddress{netaddr}
 	})
-	// TODO: confirm expected functionality before we register
-	GlobalWorkloads.RegisterBatch(krt.BatchedEventFilter(
-		func(a model.WorkloadInfo) *workloadapi.Workload {
+
+	// This is a first pass at creating the collection of services for local workloads.
+	// Here, we only include services that have a remote workload.
+	// Later, we join with all services, giving this ones here priority.
+	SplitHorizonServices := krt.NewCollection(
+		WorkloadServiceIndex.AsCollection(),
+		func(ctx krt.HandlerContext, i krt.IndexObject[string, model.WorkloadInfo]) *model.ServiceInfo {
+			localNetwork := a.Network(ctx).String()
+			svc := krt.FetchOne(ctx, a.services.Collection, krt.FilterKey(i.Key))
+			if svc == nil {
+				log.Errorf("Failed to find service %s for SplitHorizonServices", i.Key)
+				return nil
+			}
+			if svc.Scope != model.Global {
+				return nil
+			}
+			meshCfg := krt.FetchOne(ctx, a.meshConfig.AsCollection())
+			if meshCfg == nil {
+				log.Errorf("Failed to find mesh config")
+				return nil
+			}
+			sans := sets.String{}
+			for _, wl := range i.Objects {
+				if wl.Workload.Network == localNetwork {
+					continue
+				}
+				sans.Insert(spiffe.MustGenSpiffeURI(meshCfg.MeshConfig, wl.Workload.Namespace, wl.Workload.ServiceAccount))
+			}
+			if sans.IsEmpty() {
+				// Ignore workloads without remote workloads
+				return nil
+			}
+			sans = sans.Union(sets.New(svc.Service.SubjectAltNames...))
+
+			newSvc := &model.ServiceInfo{
+				Service: protomarshal.Clone(svc.Service),
+				Scope:   svc.Scope,
+			}
+			newSvc.Service.SubjectAltNames = sans.UnsortedList()
+			return precomputeServicePtr(newSvc)
+		}, opts.WithName("SplitHorizonServicesWithWorkloads")...)
+
+	// Give the SplitHorizonServicesWithWorkloads priority as they also have the sans set.
+	SplitHorizonServices = krt.JoinCollection(
```

Comment:
I'm concerned about this approach because:
1. We're duplicating services
2. There's no merge function for this join

Could we just create a new collection that depends primarily on services and fetches workloads to change the SANs?